### PR TITLE
docs(framework) Add how-to guide for designing stateful `ClientApp` objects

### PR DESCRIPTION
**Pull request by jafermarq**
Shows how to make use of the `Context`'s state to store metrics and model parameters.

Content formatted with https://github.com/dzhu/rstfmt

### CODE REVIEW

**Comment by Dawitkiros** [comment]
Hi,

I am not sure if I should create a new issue or comment here. I discussed with @jafermarq on how to implement a **stateful client** on Slack, and he gave me great tips and a link to this repository to provide my feedback on whether everything is working as expected. I followed the steps provided to store the client parameters in the context, like the following in the `fit` function right after training:

```python
# Save the parameters after training to the context for the next round
p_record = ParametersRecord()
for k, v in self.net.state_dict().items():
    # Convert to NumPy, then to Array. Add to record
    p_record[k] = ndarray_to_array(v.cpu().numpy())
# Add to context
self.client_state.parameters_records["prev_round"] = p_record
```

I loaded the context in `set_parameters` as follows:

```python
# Here, I am hoping to get the previous state dict from the client_state
prev_state_dict = {}
# Extract record from context
p_record = self.client_state.parameters_records["prev_round"]
# Deserialize arrays
for k, v in p_record.items():
    prev_state_dict[k] = torch.from_numpy(np.copy(basic_array_deserialization(v)))
prev_state_dict_list = segment_resnet_parameters(flatten_resnet_parameters(prev_state_dict))
```

Everything is working as expected, but I encountered the following warning:

```
UserWarning: The given NumPy array is not writable, and PyTorch does not support non-writable tensors. This means writing to this tensor will result in undefined behavior. You may want to copy the array to protect its data or make it writable before converting it to a tensor. This type of warning will be suppressed for the rest of this program.
```

I found a way to make it work using `np.copy()`, though I'm not sure if it's the right solution.

Thanks.

**Comment by jafermarq** [comment]
Hi @Dawitkiros , sorry for the delay. Could you try this:
```python
for k, v in p_record.items():
    # prev_state_dict[k] = torch.from_numpy(np.copy(basic_array_deserialization(v)))
    prev_state_dict[k] = torch.from_numpy(v.numpy()) # instead of the line above
```

**Comment by Dawitkiros** [comment]
Hi @jafermarq,

Your suggestion works!

**Comment by jafermarq** [comment]
> Hi @jafermarq,
> 
> Your suggestion works!

amazing! thanks for confirming. If you have some suggestions on how to improve this doc page. We are very happy to incorporate your feedback. We might be merging this PR soon.

File: doc/source/how-to-design-stateful-clients.rst [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+.. _metricsrecord: ref-api/flwr.common.MetricsRecord.html
+
+.. _numpyclient: ref-api/flwr.client.NumPyClient.html
+
+.. _parametersrecord: ref-api/flwr.common.ParametersRecord.html
+
+.. _recordset: ref-api/flwr.common.RecordSet.html#recordset
+
+By design, ClientApp_ objects are stateles. This means that the ``ClientApp`` object is
```
Comment (chongshenng) [inline doc/source/how-to-design-stateful-clients.rst:20]:
```suggestion
By design, ClientApp_ objects are stateless. This means that the ``ClientApp`` object is
```

File: doc/source/how-to-design-stateful-clients.rst [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+
+    # round 4 (note `eval_metrics` has results added in round 1&2&3)
+    configs_records={'eval_metrics': {'n_val': [2, 9, 4]}}
+    configs_records={'eval_metrics': {'n_val': [8, 2, 5]}}
+
+Saving model parameters to the context
+--------------------------------------
+
+Using ConfigsRecord_ or MetricsRecord_ to save "simple" components (e.g. list of
+`floats`, strings, booleans, etc) is fine. Flower has a specific type of record, a
```
Comment (panh99) [inline doc/source/how-to-design-stateful-clients.rst:149]:
```suggestion
Using ConfigsRecord_ or MetricsRecord_ to save "simple" components  is fine (e.g., float, integer, boolean, string, 
bytes, and lists of these types. Note that MetricsRecord_ only supports float, integer, and lists of these types) Flower has a specific type of record, a
```

File: doc/source/how-to-design-stateful-clients.rst [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+ParametersRecord_, for storing model parameters or more generally data arrays.
+
+Let's see a couple of examples of how to save NumPy arrays first and then how to save
+parameters of a PyTorch and TensorFlow model.
```
Comment (panh99) [inline doc/source/how-to-design-stateful-clients.rst:153]:
```suggestion
parameters of PyTorch and TensorFlow models.
```

File: doc/source/how-to-design-stateful-clients.rst [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+    The examples below omit the definition of a ``ClientApp`` to keep the code blocks
+    concise. To make use of ``ParametersRecord`` objects in your ``ClientApp`` you can
+    follow the same principles as outlined earlier.
+
+Saving NumPy arrays to the context
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Elements stored in a `ParametersRecord` are of type Array_, which is a data structure
+that holds ``bytes`` and metadata that can be used for deserialization. Let's see how to
+create an ``Array`` from a numpy array and insert it into a ``ParametersRecord``. Here
```
Comment (panh99) [inline doc/source/how-to-design-stateful-clients.rst:166]:
```suggestion
create an ``Array`` from a NumPy array and insert it into a ``ParametersRecord``. Here
```

File: doc/source/how-to-design-stateful-clients.rst [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+    # Let's create a simple NumPy array
+    arr_np = np.random.randn(3, 3)
+
+    # If we print it
+    # array([[-1.84242409, -1.01539537, -0.46528405],
+    #        [ 0.32991896,  0.55540414,  0.44085534],
+    #        [-0.10758364,  1.97619858, -0.37120501]])
+
+    # Now, let's serialize it an construct an Array
```
Comment (panh99) [inline doc/source/how-to-design-stateful-clients.rst:193]:
```suggestion
    # Now, let's serialize it and construct an Array
```

File: doc/source/how-to-design-stateful-clients.rst [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+
+
+    # Instantiate model as usual
+    model = Net()
+
+    # Save all elements of the state_dict into a single RecordSet
+    p_record = ParametersRecord()
+    for k, v in model.state_dict().items():
+        # Convert to NumPy, then to Array. Add to record
+        p_record[k] = array_from_numpy(v.numpy())
```
Comment (panh99) [inline doc/source/how-to-design-stateful-clients.rst:257]:
```suggestion
        p_record[k] = array_from_numpy(v.detach().cpu().numpy())
```

File: doc/source/how-to-design-stateful-clients.rst [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+
+    # Add to a context
+    context.state.parameters_records["net_parameters"] = p_record
+
+Let say now you want to apply the parameters stored in your context to a new instnace of
```
Comment (panh99) [inline doc/source/how-to-design-stateful-clients.rst:262]:
```suggestion
Let say now you want to apply the parameters stored in your context to a new instance of
```

File: doc/source/how-to-design-stateful-clients.rst [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+
+    state_dict = {}
+    # Extract record from context
+    p_record = context.state.parameters_records["net_parameters"]
+
+    # Deserialize arrays
+    for k, v in p_record.items():
+        state_dict[k] = torch.from_numpy(v.numpy())
+
+    # Applpy state dict to a new model instance
```
Comment (panh99) [inline doc/source/how-to-design-stateful-clients.rst:278]:
```suggestion
    # Apply state dict to a new model instance
```

File: doc/source/how-to-design-stateful-clients.rst [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+    model_ = Net()
+    model_.load_state_dict(state_dict)
+    # now this model has the exact same parameters as the one created earlier
+    # You can verify this by doing
+    for p, p_ in zip(model.state_dict().values(), model_.state_dict().values()):
+        assert torch.allclose(p, p_), "state_dict do not match"
```
Comment (panh99) [inline doc/source/how-to-design-stateful-clients.rst:284]:
```suggestion
        assert torch.allclose(p, p_), "`state_dict`s do not match"
```

File: doc/source/how-to-design-stateful-clients.rst [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+
+And that's it! Recall that even though this example shows how to store the entire
+`state_dict` in a ``ParametersRecord``, you can just save part of it. The process would
```
Comment (panh99) [inline doc/source/how-to-design-stateful-clients.rst:287]:
```suggestion
``state_dict`` in a ``ParametersRecord``, you can just save part of it. The process would
```

File: doc/source/how-to-design-stateful-clients.rst [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+be identical, but you might need to adjust how it is loaded into an existing model using
+PyTorch APIs.
+
+Saving Tensorflow/Keras parameters to the context
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Follow the same steps as done above but replace the `state_dict` logic with simply
```
Comment (panh99) [inline doc/source/how-to-design-stateful-clients.rst:294]:
```suggestion
Follow the same steps as done above but replace the ``state_dict`` logic with simply
```

File: doc/source/how-to-design-stateful-clients.rst [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+    def client_fn(context: Context):
+        return SimpleClient(context).to_client()  # Note we pass the context
+
+
+    # Finally, construct the clinetapp instance by means of the `client_fn` callback
+    app = ClientApp(client_fn=client_fn)
+
+If you run the app. You'll see an output similar to the one below. See how after each
+round the `n_val` entry in the context gets one additional integer ? Note that the order
+at which the `ClientApp` logs these messages might differ slightly between rounds.
```
Comment (panh99) [inline doc/source/how-to-design-stateful-clients.rst:125]:
```suggestion
in which the `ClientApp` logs these messages might differ slightly between rounds.
```

**Review by danieljanes — APPROVED** [review]

**Pull request merged**